logging: support emitting MIPI Sys-T catalog messages

dcpleung opened the pull request:
MIPI Sys-T catalog messages are similar to dictionary logging where an ID is emitted instead of the format string. This allows the format strings to be removed from the final binary to save a few bytes. This adds the necessary bits to determine to emit catalog messages when appropriate.

See individual commits for details.

dcpleung committed:
scripts: logging/dictionary: extract actual strings for db

This changes the database generation to actually extracting
individual strings instead of stuffing the whole binary sections
into the database. This allows the generation script to be
extended to accommodate more output formats.

Note that if CONFIG_LOG2_FMT_SECTION is enabled, the format
strings are in log_strings_sections, and also have associated
debug symbols in DWARF. So there is no need to manually
extract them.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: fixes pylint warnings

Fixes some easy pylint warnings.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: can specify output format

This changes the script to allow output format to be specified.
Currently, only JSON is support. This will allow supporting
other formats in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: logging/dictionary: update test for combined strings

Compilers often combine strings to conserve space, if one string is
a perfect substring of another one towards the end. So add another
string in the test to make sure dictionary logging is still working
correctly under this scenario.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
scripts: logging/dictionary: generate MIPI Sys-T collateral

This adds the ability to the dictionary logging database
generation script to output MIPI Sys-T collateral in XML.
This will allow usage of catalog logging under Sys-T, which
is similar to dictionary logging.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: syst: add support for catalog messages

MIPI Sys-T catalog messages are similar to dictionary logging
where an ID is emitted instead of the format string. This allows
the format strings to be removed from the final binary to save
a few bytes. This adds the necessary bits to determine to emit
catalog messages when appropriate.

Note that this implementation copies the argument list as-is
with string arguments stitched together since the format strings
are assumed to have been removed and they cannot be examined
to properly convert the argument lists into catalog message
payloads. Because of this, various build asserts are there to
avoid building for configurations where they are known not to
work.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
west.yml: update mipi-sys-t for catalog msg support

This updates the mipi-sys-t module to support prepared argument
list for catalog messages.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: syst: v1/imm: remove temp buffer for sync string output

The MIPI Sys-T library can now take a format string with a variable
argument list so there is no need for the temporary buffer anymore.

This saves some stack space in v1 immediate mode.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: logging/syst: update sample for combined strings

Compilers often combine strings to conserve space, if one string is
a perfect substring of another one towards the end. So add another
string in the test to make sure sys-t with catalog messages is still
working correctly under this scenario.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: syst: allow some 64-bit boards

This adds qemu_x86_64 and qemu_cortex_a53 to make sure 64-bit
support for Sys-T is not broken in the future.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
samples: logging/syst: build for catalog messages

This adds to sample.yaml to build for MIPI Sys-T catalog message
support.

Note that the current implementation of emitting Sys-T
catalog messages does not support 64-bit and architectures
having additional alignment requirements. So only allow qemu_x86
and mps2_an385 at the moment.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: v2: add kconfig to always add indexes to msg packages

This adds a new CONFIG_LOG2_MSG_PKG_ALWAYS_ADD_RO_STRING_IDXS
kconfig. If enabled, the log message packages will always have
the indexes of read-only string arguments appended to the package.
This will be selected only by those backends requiring it.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: dictionary: update package processing

With the new structure of the cbpprintf packages, the Python
parser needs to be updated to skip extra bytes before reaching
the string table.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Updated `west.yml` as https://github.com/zephyrproject-rtos/mipi-sys-t/pull/6 has been merged.

nordic-krch reviewed (commented):
Started to look into it. Wasn't yet involved much in dictionary/catalog logging.
I have 2 question:
- why string from rodata are not cataloged but appended to the log message?
- how to parse catalog mipi syst sample? I run the sample and wanted to run `log_parser.py` with xml file as database but got error.

dcpleung commented:
> * why string from rodata are not cataloged but appended to the log message?

Sys-T does not use catalog for string arguments. There is no way to tag whether a string in the argument list is a string in catalog or a full string. Therefore, each string argument needs to be replaced with the full string.

> * how to parse catalog mipi syst sample? I run the sample and wanted to run `log_parser.py` with xml file as database but got error.

There is a `systprint` utility in `modules/debug/mipi-sys-t/` that needs to be built. For example, after saving the content of a qemu_x86 run in a text file, you can do `systprint -p -c <build_dir>/zephyr/mipi_syst_collateral.xml <qemu output>` to get the messages. Note that the utility looks for lines (strictly) beginning with `SYS-T RAW DATA:` for messages to be parsed. So depending on the QEMU output, the first line may not print.

dcpleung commented on include/logging/log_msg2.h:
```diff
@@ -195,19 +195,20 @@ enum z_log_msg2_mode {
 #define Z_LOG_MSG2_STACK_CREATE(_cstr_cnt, _domain_id, _source, _level, _data, _dlen, ...) \
 do { \
 	int _plen; \
+	uint32_t flags = CBPRINTF_PACKAGE_FIRST_RO_STR_CNT(_cstr_cnt) | \
```
Updated and replaced it with a macro.

dcpleung commented on include/logging/log_msg2.h:
```diff
@@ -221,8 +222,7 @@ do { \
 	int _plen; \
-	CBPRINTF_STATIC_PACKAGE(NULL, 0, _plen, Z_LOG_MSG2_ALIGN_OFFSET, \
-				CBPRINTF_PACKAGE_FIRST_RO_STR_CNT(_cstr_cnt), \
+	CBPRINTF_STATIC_PACKAGE(NULL, 0, _plen, Z_LOG_MSG2_ALIGN_OFFSET, flags, \
```
weird... ran through twister without any issue...

nordic-krch commented:
> Therefore, each string argument needs to be replaced with the full string.

@dcpleung does it mean that strings are inserted as value into the stream and not appended like in cbprintf package? Is this somehow standardized or would it be possible to somehow send package as is (converted to mipi syst stream)? Current approach requires quite a lot of processing. Also when package is statically created then ro/rw indexes will be appended for char pointers even when they were used for `%p`. Will it not disrupt the processing?

One more thing. I didn't see `source` being sent, any plans to support it?

nordic-krch commented:
right, I only got to mipi STP specification and couldn't find anything there. Seems that i need to get into mipi_syst specification then.

dcpleung commented:
Rebased after #43520 was merged.

dcpleung committed:
samples: logging/syst: add README

Add a README.rst on how to build the MIPI Sys-T logging sample
and how to parse the output.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
CODEOWNERS: add myself for samples/subsys/logging/syst

Add myself to the CODEOWNERS for samples/subsys/logging/syst
to make checkpatch happy, and to get notifications of changes
in the sample.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Added a README for the syst sample to show how to get back readable log messages.

nordic-krch reviewed (approved):
One minor comment.

I wonder if it wouldn't be better to just dump log message as is (similar to dictionary backend) or put it into some custom mipi STP stream and then have some python script that converts this stream to mipi syst format. That would be simpler on device side, handle `%p` and ro strings could also be sent by address and fetched on the host.

MaureenHelm commented:
> Added a README for the syst sample to show how to get back readable log messages.

Thanks @dcpleung

dcpleung commented on subsys/logging/log_output_syst.c:
```diff
@@ -785,6 +784,394 @@ static int mipi_formatter(cbprintf_cb out, void *ctx,
+#endif
+
+#ifdef CONFIG_64BIT
+#define MIPI_SYST_CATMSG_ARGS_COPY MIPI_SYST_CATALOG64_ARGS_COPY
```
Hopefully we can support 64-bit in the near future.

dcpleung commented:
> I wonder if it wouldn't be better to just dump log message as is (similar to dictionary backend) or put it into some custom mipi STP stream and then have some python script that converts this stream to mipi syst format. That would be simpler on device side, handle `%p` and ro strings could also be sent by address and fetched on the host.

This is probably going to be problematic as people won't be able to use arbitrary MIPI tool to parse the log messages. Having a standard spec means multiple tools can be used (as drop-in replacement). Requiring an additional parsing steps breaks this model. I don't think people needing log messages through MIPI would want to deviate from standard.
